shared/pingpong/pingpong.go

```diff
@@ -882,7 +883,7 @@ func (pps *WorkerState) sendFromTo(
 			timeCredit -= took
 			if timeCredit > 0 {
 				time.Sleep(timeCredit)
-				timeCredit = time.Duration(0)
+				timeCredit -= time.Since(now)
```

Comment:
#3478 was removing the credits logic, which would make it worse if we were to generate > 1000 tps.

Reply:
why exactly?

Reply:
In #3478, you were calculating the next timestamp based on the current one by adding the "1/tps". This is generally correct. However, if we want to have 1000 tps, and half of the iterations takes 0.5ms while the other half takes 1.5ms, the effective TPS would be 1.5*500+1*500 = 1250 sec / 1000 txn -> 800 tps. That's why we're using the credit algorithm.

Reply:
This is not what #3478 does. I think the PR description in #3478 explains how it works.

Reply:
The description in #3478 does match the implementation.. But the goal was not to fix it for the 115 tps use case. The goal was to fix it so it would work correctly with the go 1.16 bug. If you'll take your implementation and try to use it for more than 1K tps from a single pingpong on go 1.16 it wouldn't work. With this implementation, it would.
( we need to have that in order to be able to test 10K TPS ).

Reply:
Why will my implementation not work?

Reply:
let's say we set the tps to 2000. In that case, each sleep *should* take 0.5ms. But because of this bug, it would take 1ms. That would limit the tps to 1000 ( assuming that the iteration is free ).

Reply:
Btw - I do like the simplicity in your implementation, but I think we should work a bit more on it. Let's rebase your branch based on this one, and conduct some further testing against high-volume TPS ? I want to see that it would work correctly both in the cases of random iteration timing as well as high TPS.

Reply:
> let's say we set the tps to 2000. In that case, each sleep _should_ take 0.5ms. But because of this bug, it would take 1ms. That would limit the tps to 1000 ( assuming that the iteration is free ).

No, because some sleeps will be skipped. But I'm definitely not against using nanosleep for more precision.

> Btw - I do like the simplicity in your implementation, but I think we should work a bit more on it. Let's rebase your branch based on this one, and conduct some further testing against high-volume TPS ? I want to see that it would work correctly both in the cases of random iteration timing as well as high TPS.

Sure!